packages/@aws-cdk/aws-route53/lib/vpc-endpoint-service-domain-name.ts

```diff
@@ -0,0 +1,197 @@
+    super(scope, id);
+
+    // Make sure a user doesn't accidentally add multiple domains
+    this.validateProps(props);
+    VpcEndpointServiceDomainName.endpointServices.push(props.endpointService);
+
+    const serviceId = props.endpointService.vpcEndpointServiceId;
+    const privateDnsName = props.domainName;
+
+    // Enable Private DNS on the endpoint service and retrieve the AWS-generated configuration
+    const privateDnsConfiguration = this.getPrivateDnsConfiguration(id, serviceId, privateDnsName);
+
+    // Tell AWS to verify that this account owns the domain attached to the service
+    this.verifyPrivateDnsConfiguration(privateDnsConfiguration, props.publicHostedZone);
+
+    // Finally, don't do any of the above before the endpoint service is created
+    this.node.addDependency(props.endpointService);
+  }
+
+  private validateProps(props: VpcEndpointServiceDomainNameProps): void {
+    if (VpcEndpointServiceDomainName.endpointServices.includes(props.endpointService)) {
+      throw new Error(
+        `Cannot create a VpcEndpointServiceDomainName for service ${props.endpointService.node.uniqueId}, another VpcEndpointServiceDomainName is already associated with it`);
+    }
+  }
+
+  /**
+   * Sets up Custom Resources to make AWS calls to set up Private DNS on an endpoint service,
+   * returning the values to use in a TxtRecord, which AWS uses to verify domain ownership.
+   */
+  private getPrivateDnsConfiguration(id: string, serviceId: string, privateDnsName: string): PrivateDnsConfiguration {
+
+    // The custom resource which tells AWS to enable Private DNS on the given service, using the given domain name
+    // AWS will generate a name/value pair for use in a TxtRecord, which is used to verify domain ownership.
+    const enablePrivateDnsAction = {
+      service: 'EC2',
+      action: 'modifyVpcEndpointServiceConfiguration',
+      parameters: {
+        ServiceId: serviceId,
+        PrivateDnsName: privateDnsName,
+      },
+      physicalResourceId: PhysicalResourceId.of(id),
+    };
+    const removePrivateDnsAction = {
+      service: 'EC2',
+      action: 'modifyVpcEndpointServiceConfiguration',
+      parameters: {
+        ServiceId: serviceId,
+        RemovePrivateDnsName: true,
+      },
+    };
+    const enable = new AwsCustomResource(this, 'EnableDns', {
+      onCreate: enablePrivateDnsAction,
+      onUpdate: enablePrivateDnsAction,
+      onDelete: removePrivateDnsAction,
+      policy: AwsCustomResourcePolicy.fromSdkCalls({
+        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
+      }),
+    });
+
+    // Look up the name/value pair if the domain changes, or the service changes,
+    // which would cause the values to be different. If the unique ID changes,
+    // the resource may be entirely recreated, so we will need to look it up again.
+    const lookup = hashcode(Names.uniqueId(this) + serviceId + privateDnsName);
```

Comment:
If you want to trigger a re-lookup, you shouldn't change the returned `physicalResourceId`, but you should be changing the parameters to the custom resource (or its logical ID).

I guess doing this might accidentally work since `physicalResourceId` is a parameter, but it's a parameter with semantics that I'm not too sure about changing. I think I'd feel better if you added a dummy input parameter to the `AwsCustomResource` which only exists to trigger an update.

(Maybe this is worth a special subclass even, because that kind of behavior only makes sense for a read action... and you also wouldn't have to specify both `onCreate` and `onUpdate`).

Reply:
Yes, `serviceId` is a Ref which would be a token. Updating to use the unique Id of the service instead